File: auto-mpg.py
import pandas as pd
import numpy as np
from sklearn.impute import KNNImputer
from matplotlib import pyplot as plt
from scipy.interpolate import make_interp_spline


df = pd.read_csv('./auto-mpg.csv')
df = df.replace('?',0)
knn = KNNImputer(missing_values=0, n_neighbors=5)
df.horsepower = knn.fit_transform(df.horsepower.values.reshape(-1,1))
mean_df = df.groupby('model year').mean().reset_index()

B_spline_coeff = make_interp_spline(mean_df['model year'].values, mean_df.horsepower.values)
X_Final = np.linspace(mean_df['model year'].values.min(), mean_df['model year'].values.max(), 500)
Y_horse = B_spline_coeff(X_Final)

B_spline_coeff = make_interp_spline(mean_df['model year'].values, mean_df.displacement.values)
Y_displacement = B_spline_coeff(X_Final)


plt.rcParams['figure.figsize'] = (7.5,4)
fig, ax = plt.subplots()
fig.set_facecolor('#eaf0f6')
ax.spines[['top','right']].set_visible(False)
ax.set_facecolor('#eaf0f6')
ax.grid(axis='y')
ax.plot(X_Final,Y_horse,color='#4285F4',label='Engine Displacement')
ax.plot(X_Final,Y_displacement,color='#ea4335',label='Horsepower')
ax.scatter(mean_df['model year'].values,mean_df.horsepower.values,color='#4285F4')
ax.scatter(mean_df['model year'].values,mean_df.displacement.values,color='#ea4335')
ax.set_yticks([100,200,300])
ax.set_xlabel('Model Year')
plt.legend(bbox_to_anchor =(0.75, 1.15), ncol = 2,frameon=False,)
plt.show()

File: auto-mpg.r
library(readr)
library(ggplot2)

auto_mpg <- read_csv("auto-mpg.csv")

auto_mpg$horsepower <- replace(auto_mpg$horsepower,auto_mpg$horsepower=="?",0)

preProcValues <- preProcess(auto_mpg[,c('horsepower')],
                            method = c("knnImpute"),
                            k = 20,
                            knnSummary = mean)
impute_cars_info <- predict(preProcValues, auto_mpg)
impute_cars_info$horsepower <- strtoi(impute_cars_info$horsepower)

agg_df <- aggregate(impute_cars_info, by=list(impute_cars_info$`model year`), FUN=mean)
plot(agg_df$Group.1,agg_df$displacement, type = "o", col = "red")
points(agg_df$Group.1,agg_df$horsepower,type="o", col="blue")



ggplot(agg_df, aes(x=Group.1)) + 
  geom_line(aes(y = displacement), color = "#ea4335") + 
  geom_line(aes(y = horsepower), color="#4285F4") +
  geom_point(aes(y = displacement), color = "#ea4335")+
  geom_point(aes(y = horsepower), color = "#4285F4")+
  labs(x = "Model Year", y="") +
  theme(legend.position = "top") +
  coord_cartesian(ylim = c(0, 300)) +
  theme_classic()+
  scale_x_discrete(limits = c(70, 72, 74,76, 78, 80, 82))


                  
                  